File: framework/fit/python/fit_flowable/subscription/abstract_subscription.py
# -- encoding: utf-8 --
# Copyright (c) 2024 Huawei Technologies Co., Ltd. All Rights Reserved.
# This file is a part of the ModelEngine Project.
# Licensed under the MIT License. See License.txt in the project root for license information.
# ======================================================================================================================
"""
功 能：订阅关系的抽象父类。
"""
from abc import abstractmethod

from fit_flowable.core.subscription import Subscription
from fit_flowable.util.concurrent import AtomicBool


class AbstractSubscription(Subscription):
    """
    订阅关系的抽象父类。
    """

    def __init__(self):
        self._cancelled: AtomicBool = AtomicBool()

    def request(self, count: int) -> None:
        if count <= 0:
            raise ValueError(f"The number of elements to request must be positive. [count={count}]")
        if self._cancelled.get():
            return
        self.do_request(count)

    @abstractmethod
    def do_request(self, count) -> None:
        """
        请求指定数量的数据时的具体行为，仅当前请求没有被取消时有效。

        :param count: 请求的数据的数量
        """
        pass

    def cancel(self) -> None:
        if not self._cancelled.compare_and_set(False, True):
            return
        self.do_cancel()

    def do_cancel(self) -> None:
        """
        取消当前的订阅关系的具体行为，仅当前请求没有被取消时有效，默认为空。
        """
        pass

    def get_cancelled(self) -> bool:
        return self._cancelled.get()

File: framework/fit/java/fit-reactor/src/main/java/com/huawei/fitframework/flowable/subscription/AbstractSubscription.java
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2024-2024. All rights reserved.
 */

package com.huawei.fitframework.flowable.subscription;

import static com.huawei.fitframework.inspection.Validation.greaterThan;

import com.huawei.fitframework.flowable.Subscription;

import java.util.concurrent.atomic.AtomicBoolean;

/**
 * 表示 {@link Subscription} 的抽象实现类。
 *
 * @author 季聿阶
 * @since 2024-02-11
 */
public abstract class AbstractSubscription implements Subscription {
    private final AtomicBoolean cancelled = new AtomicBoolean();

    @Override
    public void request(long count) {
        greaterThan(count, 0, "The number of elements to request must be positive. [count={0}]", count);
        if (this.cancelled.get()) {
            return;
        }
        this.request0(count);
    }

    /**
     * 请求指定数量的数据。
     * <p>当前请求没有被取消。</p>
     *
     * @param count 表示请求的数据的数量的 {@code long}。
     */
    protected abstract void request0(long count);

    @Override
    public void cancel() {
        if (!this.cancelled.compareAndSet(false, true)) {
            return;
        }
        this.cancel0();
    }

    /**
     * 取消当前的订阅关系。
     * <p>默认为空。</p>
     */
    protected void cancel0() {}

    @Override
    public boolean isCancelled() {
        return this.cancelled.get();
    }
}
